README
NAME

    PerlIO::via::QuotedPrint - PerlIO layer for quoted-printable strings

SYNOPSIS

    use PerlIO::via::QuotedPrint;

    open(my $in, '<:via(QuotedPrint)', 'file.qp') or
        die "Can't open file.qp for reading: $!\n";

    open(my $out, '>:via(QuotedPrint)', 'file.qp') or
        die "Can't open file.qp for writing: $!\n";

DESCRIPTION

    This module implements a PerlIO layer that works on files encoded in the
    quoted-printable format. It will decode from quoted-printable while reading
    from a handle, and it will encode as quoted-printable while writing to a
    handle.

INSTALLATION

    See the INSTALL file.

COPYRIGHT

    Copyright (C) 2002-2004, 2012 Elizabeth Mattijsen. All rights reserved.
    Copyright (C) 2015, 2020, 2022 Steve Hay. All rights reserved.

LICENCE

    This distribution is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, i.e. under the terms of either the GNU
    General Public License or the Artistic License, as specified in the LICENCE
    file.


Changes
Revision history for Perl distribution PerlIO-via-QuotedPrint.

0.11 Development

    - TODO

0.10 2022-05-22

    - Add missing exit() to t/QuotedPrint.t.  [Karl Williamson, PR#1]

0.09 2020-12-08

    - Added INSTALLDIRS, accounting for the @INC reordering change in Perl 5.12.
      See Perl RT#116479 for details.  (PerlIO-via-QuotedPrint entered the perl
      core in Perl 5.8.0 so that's what the lower bound of the check should
      strictly be, but since we only support Perl 5.8.1 and higher anyway we do
      not need to check the lower bound.  The upper bound is correctly Perl
      5.11.0 since the @INC reordering change in question (Perl core commit
      #b9ba2fadb1) first appeared in Perl 5.11.0.)

    - Added optional Changes testing (skipped unless AUTHOR_TESTING).

    - Reformatted Changes file as per CPAN::Changes::Spec.

    - Added optional POD coverage testing (skipped unless AUTHOR_TESTING).

    - Added optional POD testing (skipped unless AUTHOR_TESTING).

    - Added optional Perl::Critic testing (skipped unless AUTHOR_TESTING).

    - Made code Perl::Critic clean.

    - Included GitHub repository URLs in metadata now that source code has been
      uploaded to GitHub.

    - Set minimum required ExtUtils::MakeMaker version to 6.64 to ensure that
      all parameters used are supported, to save jumping through hoops to
      support earlier versions.  (This should not be a problem since
      ExtUtils::MakeMaker 6.64 is easily installed into Perl 5.8.1 and above,
      that being the whole point of the new choice of minimum supported Perl
      version.)

    - Set minimum required Perl version to 5.8.1.  This is in line with the
      minimum requirement of the "Perl Toolchain".

0.08 2015-05-18

    - Synchronized t/QuotedPrint.t with the core perl version.

0.07 2012-04-05

    - Brought up to date with current code esthetics.  Verified it runs on
      5.14.2, threaded and unthreaded.

    - Made sure the test-suite cleans up on multi-versioned systems.

0.06 2003-12-28

    - Added automatic required modules update using Devel::Required.

0.05 2003-08-13

    - Cleaned up Makefile.PL and updated copyright info.  Made sure the
      test-suite runs with strict and warnings enabled.

0.04 2002-07-11

    - Adapted Perl core magic in test-suite to conform to Jarkko Hietaniemi's
      official magic.

    - Rearranged top of source to not need fully qualified VERSION anymore.

    - Checked for defined() problems: none found.

0.03 2002-07-10

    - Added Perl CORE magic pragma to test-suite.  Changed name of test-suite to
      QuotedPrint.t to facilitate inclusion in Perl core.

    - Migrated to the PerlIO::via:: namespace.

0.02 2002-07-09

    - Fixed some documentation nits reported by Rafael Garcia-Suarez.

    - Version number now a '' string (oops).

    - Changed object to glob reference instead of list reference.  Should be a
      little memory saving.

0.01 2002-07-01

    - Migrated from the now obsoleted IOLayer:: namespace.


lib/PerlIO/via/QuotedPrint.pm
# Copyright (C) 2002-2004, 2012 Elizabeth Mattijsen.  All rights reserved.
# Copyright (C) 2015 Steve Hay.  All rights reserved.

# This module is free software; you can redistribute it and/or modify it under
# the same terms as Perl itself, i.e. under the terms of either the GNU General
# Public License or the Artistic License, as specified in the F<LICENCE> file.

package PerlIO::via::QuotedPrint;

use 5.008001;

# be as strict as possible
use strict;

our $VERSION = '0.11';

# modules that we need
use MIME::QuotedPrint (); # no need to pollute this namespace

# satisfy -require-
1;

#-------------------------------------------------------------------------------
#
# Standard Perl features
#
#-------------------------------------------------------------------------------
#  IN: 1 class to bless with
#      2 mode string (ignored)
#      3 file handle of PerlIO layer below (ignored)
# OUT: 1 blessed object

sub PUSHED { bless \*PUSHED,$_[0] } #PUSHED

#-------------------------------------------------------------------------------
#  IN: 1 instantiated object (ignored)
#      2 handle to read from
# OUT: 1 decoded string

sub FILL {

    # decode and return
    my $line= readline( $_[1] );
    return ( defined $line )
      ? MIME::QuotedPrint::decode_qp($line)
      : undef;
} #FILL

#-------------------------------------------------------------------------------
#  IN: 1 instantiated object (ignored)
#      2 buffer to be written
#      3 handle to write to
# OUT: 1 number of bytes written

sub WRITE {

    # encode and write to handle: indicate result
    return ( print { $_[2] } MIME::QuotedPrint::encode_qp( $_[1] ) )
      ? length( $_[1] )
      : -1;
} #WRITE

#-------------------------------------------------------------------------------

__END__

=head1 NAME

PerlIO::via::QuotedPrint - PerlIO layer for quoted-printable strings

=head1 SYNOPSIS

    use PerlIO::via::QuotedPrint;

    open(my $in, '<:via(QuotedPrint)', 'file.qp') or
        die "Can't open file.qp for reading: $!\n";

    open(my $out, '>:via(QuotedPrint)', 'file.qp') or
        die "Can't open file.qp for writing: $!\n";

=head1 DESCRIPTION

This module implements a PerlIO layer that works on files encoded in the
quoted-printable format.  It will decode from quoted-printable while reading
from a handle, and it will encode as quoted-printable while writing to a handle.

=head1 EXPORTS

I<None>.

=head1 KNOWN BUGS

I<None>.

=head1 FEEDBACK

Patches, bug reports, suggestions or any other feedback is welcome.

Patches can be sent as GitHub pull requests at
L<https://github.com/steve-m-hay/PerlIO-via-QuotedPrint/pulls>.

Bug reports and suggestions can be made on the CPAN Request Tracker at
L<https://rt.cpan.org/Public/Bug/Report.html?Queue=PerlIO-via-QuotedPrint>.

Currently active requests on the CPAN Request Tracker can be viewed at
L<https://rt.cpan.org/Public/Dist/Display.html?Status=Active;Queue=PerlIO-via-QuotedPrint>.

Please test this distribution.  See CPAN Testers Reports at
L<https://www.cpantesters.org/> for details of how to get involved.

Previous test results on CPAN Testers Reports can be viewed at
L<https://www.cpantesters.org/distro/P/PerlIO-via-QuotedPrint.html>.

Please rate this distribution on CPAN Ratings at
L<https://cpanratings.perl.org/rate/?distribution=PerlIO-via-QuotedPrint>.

=head1 SEE ALSO

L<PerlIO::via>,
L<MIME::QuotedPrint>.

=head1 ACKNOWLEDGEMENTS

Based on an example in the standard library module MIME::QuotedPrint in Perl
(version 5.8.0).

=head1 AVAILABILITY

The latest version of this module is available from CPAN (see
L<perlmodlib/"CPAN"> for details) at

L<https://metacpan.org/release/PerlIO-via-QuotedPrint> or

L<https://www.cpan.org/authors/id/S/SH/SHAY/> or

L<https://www.cpan.org/modules/by-module/PerlIO/>.

The latest source code is available from GitHub at
L<https://github.com/steve-m-hay/PerlIO-via-QuotedPrint>.

=head1 INSTALLATION

See the F<INSTALL> file.

=head1 AUTHOR

Elizabeth Mattijsen E<lt>L<liz@dijkmat.nl|mailto:liz@dijkmat.nl>E<gt>.

Steve Hay E<lt>L<shay@cpan.org|mailto:shay@cpan.org>E<gt> is now maintaining
PerlIO::via::QuotedPrint as of version 0.08.

=head1 COPYRIGHT

Copyright (C) 2002-2004, 2012 Elizabeth Mattijsen.  All rights reserved.

Copyright (C) 2015, 2020 Steve Hay.  All rights reserved.

=head1 LICENCE

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself, i.e. under the terms of either the GNU General
Public License or the Artistic License, as specified in the F<LICENCE> file.

=head1 VERSION

Version 0.11

=head1 DATE

TODO

=head1 HISTORY

See the F<Changes> file.

=cut


Makefile.PL
#!perl
#===============================================================================
#
# Makefile.PL
#
# DESCRIPTION
#   Makefile creation script.
#
# COPYRIGHT
#   Copyright (C) 2015 Steve Hay.  All rights reserved.
#
# LICENCE
#   This script is free software; you can redistribute it and/or modify it under
#   the same terms as Perl itself, i.e. under the terms of either the GNU
#   General Public License or the Artistic License, as specified in the LICENCE
#   file.
#
#===============================================================================

use 5.008001;

use strict;
use warnings;

use Config qw(%Config);
use ExtUtils::MakeMaker 6.64;
use ExtUtils::MakeMaker qw(WriteMakefile);

#===============================================================================
# INITIALIZATION
#===============================================================================

BEGIN {
    unless ($Config{useperlio}) {
        die "OS unsupported: This module requires a perlio-enabled perl.\n";
    }
}

#===============================================================================
# MAIN PROGRAM
#===============================================================================

MAIN: {
    WriteMakefile(
        NAME          => 'PerlIO::via::QuotedPrint',
        ABSTRACT_FROM => 'lib/PerlIO/via/QuotedPrint.pm',
        AUTHOR        => 'Elizabeth Mattijsen <liz@dijkmat.nl>, Steve Hay <shay@cpan.org>',
        LICENSE       => 'perl_5',
        VERSION_FROM  => 'lib/PerlIO/via/QuotedPrint.pm',

        META_MERGE => {
            'meta-spec' => {
                version => 2
            },

            resources => {
                repository => {
                    type => 'git',
                    url  => 'https://github.com/steve-m-hay/PerlIO-via-QuotedPrint.git'
                }
            },

            optional_features => {
                changestest => {
                    description => 'Changes testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::CPAN::Changes' => '0'
                            }
                        }
                    }
                },

                critictest => {
                    description => 'Perl::Critic testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Perl::Critic' => '0'
                            }
                        }
                    }
                },

                podtest => {
                    description => 'POD testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod' => '1.00'
                            }
                        }
                    }
                },

                podcoveragetest => {
                    description => 'POD coverage testing',
                    prereqs => {
                        test => {
                            requires => {
                                'Test::Pod::Coverage' => '0.08'
                            }
                        }
                    }
                }
            }
        },

        MIN_PERL_VERSION => '5.008001',

        CONFIGURE_REQUIRES => {
            'Config'              => '0',
            'ExtUtils::MakeMaker' => '6.64',
            'perl'                => '5.008001',
            'strict'              => '0',
            'warnings'            => '0'
        },

        TEST_REQUIRES => {
            'Test::More' => '0'
        },

        PREREQ_PM => {
            'MIME::QuotedPrint' => '0',
            'strict'            => '0'
        },

        INSTALLDIRS => ($] < 5.011 ? 'perl' : 'site'),

        dist => {
            PREOP   => 'find $(DISTVNAME) -type d -print|xargs chmod 0755 && ' .
                       'find $(DISTVNAME) -type f -print|xargs chmod 0644',
            TO_UNIX => 'find $(DISTVNAME) -type f -print|xargs dos2unix'
        }
    );
}

#===============================================================================
